ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingValidatingCommitter.scala

```diff
@@ -0,0 +1,100 @@
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{DamlStateKey, DamlStateValue}
+import com.daml.ledger.participant.state.kvutils.{Bytes, Fingerprint}
+import com.daml.ledger.participant.state.v1.{ParticipantId, SubmissionResult}
+import com.daml.ledger.validator.LedgerStateOperations.Value
+import com.daml.ledger.validator.SubmissionValidator.RawKeyValuePairs
+import com.daml.ledger.validator.caching.{
+  CacheUpdatePolicy,
+  CachingDamlLedgerStateReaderWithFingerprints
+}
+import com.daml.ledger.validator.{
+  LedgerStateAccess,
+  StateAccessingValidatingCommitter,
+  StateKeySerializationStrategy
+}
+import com.daml.timer.RetryStrategy
+
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+/**
+  * A pre-executing validating committer based on [[LedgerStateAccess]] (that does not provide fingerprints
+  * alongside values), parametric in the logic that produces a fingerprint given a value.
+  *
+  * @param now The record time provider.
+  * @param keySerializationStrategy The key serializer used for state keys.
+  * @param validator The pre-execution validator.
+  * @param valueToFingerprint The logic that produces a fingerprint given a value.
+  * @param postExecutionFinalizer The post-execution finalizer that will also perform conflicts detection and
+  *                               time bounds checks.
+  * @param stateValueCache The cache instance for state values.
+  * @param cacheUpdatePolicy The caching policy for values.
+  * @param materializer The Akka materializer.
+  * @tparam LogResult type of the offset used for a log entry.
+  */
+class PreExecutingValidatingCommitter[LogResult](
+    now: () => Instant,
+    keySerializationStrategy: StateKeySerializationStrategy,
+    validator: PreExecutingSubmissionValidator[RawKeyValuePairs],
+    valueToFingerprint: Option[Value] => Fingerprint,
+    postExecutionFinalizer: PostExecutionFinalizerWithFingerprintsFromValues[LogResult],
+    stateValueCache: Cache[DamlStateKey, (DamlStateValue, Fingerprint)],
+    cacheUpdatePolicy: CacheUpdatePolicy)(implicit materializer: Materializer)
+    extends StateAccessingValidatingCommitter[LogResult] {
+
+  /**
+    * Pre-executes and then finalizes a submission.
+    */
+  override def commit(
+      correlationId: String,
+      submissionEnvelope: Bytes,
+      submittingParticipantId: ParticipantId,
+      ledgerStateAccess: LedgerStateAccess[LogResult])(
+      implicit executionContext: ExecutionContext): Future[SubmissionResult] =
+    // Fidelity level 1: sequential pre-execution. Implemented as: the pre-post-exec pipeline is a single transaction.
+    ledgerStateAccess.inTransaction { ledgerStateOperations =>
+      for {
+        preExecutionOutput <- validator
+          .validate(
+            submissionEnvelope,
+            correlationId,
+            submittingParticipantId,
+            CachingDamlLedgerStateReaderWithFingerprints(
```

Comment:
For pre-execution we shouldn't need new instances of `LedgerStateReader` for every request.

Reply:
We do, but only because the transaction is opened every time a commit operation is initiated (and only in the context of the transaction we have access to the operations).